tests/transports/test_all_plugins.py

```diff
@@ -1365,3 +1279,73 @@ def test_exec_with_wrong_stdin(self, custom_transport):
         with custom_transport as transport:
             with self.assertRaises(ValueError):
                 transport.exec_command_wait('cat', stdin=1)
+
+
+class TestDirectScheduler(unittest.TestCase):
+    """
+    Test how the direct scheduler works.
+
+    While this is technically a scheduler test, I put it under the transport tests
+    because 1) in reality I am testing the interaction of each transport with the
+    direct scheduler; 2) the direct scheduler is always available; 3) I am reusing
+    the infrastructure to test on multiple transport plugins.
+    """
+
+    @run_for_all_plugins
+    def test_asynchronous_execution(self, custom_transport):
+        """Test that the execution of a long(ish) command via the direct scheduler does not block.
+
+        This is a regression test for #3094, where running a long job on the direct scheduler
+        (via SSH) would lock the interpreter until the job was done.
+        """
+        # Use a unique name, using the current process PID as well, to avoid concurrent tests (or very rapid
+        # tests that follow each other) to overwrite the same destination
+        script_fname = f'sleep-submit-{os.getpid()}-{custom_transport.__class__.__name__}.sh'
+
+        scheduler = SchedulerFactory('direct')()
+        scheduler.set_transport(custom_transport)
+        with custom_transport as transport:
+            with tempfile.NamedTemporaryFile() as tmpf:
+                # Put a submission script that sleeps 10 seconds
+                tmpf.write(b'#!/bin/bash\nsleep 10\n')
+                tmpf.flush()
+
+                transport.chdir('/tmp')
```

Comment:
This is `localhost` here, right?

Instead of the file transfer, it could also just use the temporary file directly.
This would also take care of removing the file in case an exception is encountered.

Reply:
Yes, but I've preferred to keep it ready to be extended to different machines (I also now remove the file via the transport in the finally block.
I still use `os.kill` for simplicity, though. It should be easy to just change that if we decide to extend testing to further machines (there are other tests that need to be adapted as well) but at least the changes to apply in that case remain confined.